File: backend/app/core/upgrades/stat_modifiers/knockback_1.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class Knockback1(UpgradeImplementation):
    id = 'knockback_1'
    name = 'Knockback Boost'
    description = '+5% knockback'
    rarity = 'common'
    type = 'stat_modifier'
    cost = 2

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 0.05
        appliers.modifiers.add_multiplier('attack', 'knockback', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/stat_modifiers/knockback_1.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const Knockback1Def: UpgradeDef = {
  id: "knockback_1",
  name: "Knockback Boost",
  description: "+5% knockback",
  rarity: RarityID.Common,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 2,
  targetClass: UpgradeTargetID.Attack,
  fieldInTargetClass: UpgradeStatID.Knockback,
  value: 0.05,
  isMultiplier: true,
  stackable: true,
  maxStacks: 5,
}

export class Knockback1 extends Upgrade {}
